### PR TITLE
*: Prevent resolved-ts from being blocked by pipelined DML transactions.

### PR DESCRIPTION
**Pull request by ekexium**
### What is changed and how it works?

Issue Number: ref #17459

What's Changed:

```commit-message
Prevent resolved-ts from being blocked by pipelined DML transactions.
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: components/resolved_ts/src/resolver.rs [hunk 1]
```diff
@@ -710,4 +859,31 @@ mod tests {
...
+
+        // track a large txn lock
+        resolver.track_lock(1.into(), key.clone(), None, 1).unwrap();
+        assert_eq!(resolver.num_locks(), 1);
+        assert_eq!(resolver.num_transactions(), 1);
+        assert_eq!(resolver.locks_by_key.len(), 0);
+        assert_eq!(resolver.large_txn_ts.len(), 1);
+        assert_eq!(resolver.large_txn_key_representative.len(), 1);
+        assert_eq!(resolver.resolved_ts(), TimeStamp::zero());
+
```
Comment (you06) [inline components/resolved_ts/src/resolver.rs:878]: May also test the large txn who has the min `start_ts` without min `min_commit_ts`.

```
txn1: start_ts: 1, min_commit_ts: 10
txn2: start_ts: 2, min_commit_ts: 5
=>
resolve_ts == 5
```

File: src/storage/txn/commands/txn_heart_beat.rs [hunk 2]
```diff
@@ -71,10 +73,28 @@ impl<S: Snapshot, L: LockManager> WriteCommand<S, L> for TxnHeartBeat {
...
+                }
+
+                // only for non-async-commit pipelined transactions, we can update the
+                // min_commit_ts
+                if !lock.use_async_commit
+                    && lock.generation > 0
+                    && self.min_commit_ts > 0
+                    && lock.min_commit_ts < self.min_commit_ts.into()
+                {
+                    lock.min_commit_ts = self.min_commit_ts.into();
```
Comment (you06) [inline src/storage/txn/commands/txn_heart_beat.rs:90]: What will be if the `lock.min_commit_ts` is not same with `min_commit_ts` in txn cache? I think the `min_commit_ts` txn cache is always less or equal than `lock.min_commit_ts`, am I right?

**Review by you06 — COMMENTED** [review]
rest LGTM

File: components/resolved_ts/src/resolver.rs [hunk 3]
```diff
@@ -80,6 +81,13 @@ pub struct Resolver {
     locks_by_key: HashMap<Arc<[u8]>, TimeStamp>,
     // start_ts -> locked keys.
     lock_ts_heap: BTreeMap<TimeStamp, TxnLocks>,
+    // the start_ts of large transactions, which use a different tracking strategy with normal
+    // transactions.
+    large_txn_ts: HashSet<TimeStamp>,
+    // each large transaction tracked by this resolver has a representative key tracked. So that
+    // when the large transaction is rolled back, we can rely on this key to guarantee that
+    // eventually there will be no dangling locks.
+    large_txn_key_representative: HashMap<Arc<[u8]>, TimeStamp>,
```
Comment (cfzjywxk) [inline components/resolved_ts/src/resolver.rs:90]: > we can rely on this key to guarantee that eventually there will be no dangling locks.

From the code it used to clear the `large_txn_ts` when related locks are untracked, what is the relationship with dangling?


Could the `HashMap<Key, TimeStamp>,` be used instead as there would not be too many representative keys from large transactions?

File: src/server/service/kv.rs [hunk 4]
```diff
@@ -2265,6 +2275,32 @@ fn future_raw_coprocessor<E: Engine, L: LockManager, F: KvFormat>(
...
+                txn_status.commit_ts.into(),
+                txn_status.rolled_back,
+            );
+            if txn_status.is_completed {
+                // large_txn_cache is only for **ongoing** large txns, so remove it when
+                // completed.
+                assert_matches!(txn_state, TxnState::Committed { .. } | TxnState::RolledBack);
+                cache.remove_large_txn(txn_status.start_ts.into());
+            }
+            cache.upsert(txn_status.start_ts.into(), txn_state, now);
```
Comment (cfzjywxk) [inline src/server/service/kv.rs:2298]: The cache is possible blocking operation, this task should be handled by the scheduler worker pool and avoid its impact on the gRPC threads.

File: components/resolved_ts/src/resolver.rs [hunk 5]
```diff
@@ -429,45 +482,141 @@ impl Resolver {
...
 
     pub(crate) fn take_last_attempt(&mut self) -> Option<LastAttempt> {
         self.last_attempt.take()
     }
+
+    fn track_large_txn_lock(
+        &mut self,
+        start_ts: TimeStamp,
+        key: Vec<u8>,
+    ) -> Result<(), MemoryQuotaExceeded> {
```
[the remark this answers is not in the document]
Comment (ekexium) [reply components/resolved_ts/src/resolver.rs:606]: While the current implementation is unlikely to consume significant memory, I've opted to maintain the function signature for consistency and to facilitate potential future enhancements.

File: src/storage/txn/commands/txn_heart_beat.rs (hunk 2 above)
Comment (ekexium) [reply src/storage/txn/commands/txn_heart_beat.rs:90]: Yes. `min_commit_ts` in the lock should be the source of truth in tikv side. txn_status_cache only caches a "recent enough" `min_commit_ts`, but does not guarantee its consistency with that in the lock.

File: components/resolved_ts/src/resolver.rs (hunk 3 above)
Comment (ekexium) [reply components/resolved_ts/src/resolver.rs:90]: > what is the relationship with dangling?

"orphaned txns" might be more accurate. This is for the case that a large txn entry in resolver is never removed.

> Could the HashMap<Key, TimeStamp>, be used instead

Makes sense. I think `Vec<u8>` is more appropriate here as `Key` is more often for encoded keys.

File: src/server/service/kv.rs (hunk 4 above)
Comment (ekexium) [reply src/server/service/kv.rs]: Good catch

**Review by you06 — APPROVED** [review]

**Pull request merged**